### PR TITLE
[REVIEW] Fixing loc ordering issue in dataframe

### LINKED ISSUE #4265: [BUG] Using df.loc[idx] returns rows out of order.
**Describe the bug**
Sometimes when you use loc to select specific rows (i.e. df.loc[idx]), the rows are returned out of order with relation to the ordering of idx.

**Steps/Code to reproduce bug**
df = cudf.DataFrame()
df['a'] = np.array([2987000., 2987001., 2987002., 2987003., 2987004., 2987005.,
       2987006., 2987007., 2987008., 2987009., 2987010., 2987011.,
       2987012., 2987013., 2987014., 2987015., 2987016., 2987017.,
       2987018., 2987019., 2987020., 2987021., 2987022., 2987023.,
       2987024., 2987025., 2987026., 2987027., 2987028., 2987029.,
       2987030., 2987031., 2987032., 2987033., 2987034., 2987035.,
       2987036., 2987037., 2987038., 2987039., 2987040., 2987041.,
       2987042., 2987043., 2987044., 2987045., 2987046., 2987047.,
       2987048., 2987049., 2987050., 2987051., 2987052., 2987053.,
       2987054., 2987055., 2987056., 2987057., 2987058., 2987059.,
       2987060., 2987061., 2987062., 2987063., 2987064., 2987065.,
       2987066., 2987067., 2987068., 2987069., 2987070., 2987071.,
       2987072., 2987073., 2987074., 2987075., 2987076., 2987077.,
       2987078., 2987079., 2987080., 2987081., 2987082., 2987083.,
       2987084., 2987085., 2987086., 2987087., 2987088., 2987089.,
       2987090., 2987091., 2987092., 2987093., 2987094., 2987095.,
       2987096., 2987097., 2987098., 2987099.], dtype='float32')
df['b'] = np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype='int8')
df = df.set_index('a')
split = 3*len(df)//4
idxA = df.index[:split]
idxB = df.index[split:]
print( df.loc[idxA] )

**Expected behavior**
The first row printed should be row with index 2987000. but it is not. In the Jupyter notebook here:
https://github.com/cdeotte/RAPIDS-development/blob/master/bugs/bug002.ipynb
we see that the first row printed has index 2987064.0

**Environment overview (please complete the following information)**
The stable 0.12.0 install of RAPIDS

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
As join doesn't preserve the ordering, the order was getting jumbled which created this issue.
close #4265

### CODE REVIEW

**Comment by jrhemstad** [comment]
`loc` seems like a common operation. Having to do a sort every time is very expensive.

**Comment by rgsl888prabhu** [comment]
> `loc` seems like a common operation. Having to do a sort every time is very expensive.

This applies to only if key provided is a array like object and I am not sure about any less expensive alternative.

**Comment by kkraus14** [comment]
> This applies to only if key provided is a array like object and I am not sure about any less expensive alternative.

It was on my todo list but I dropped the ball. Could you raise an issue with a feature request for libcudf support for something to accomplish the behavior of `loc`? This doesn't include index columns, but just generating a gather map based on values in a specified set of columns.

File: python/cudf/cudf/core/indexing.py [hunk 1]
```diff
@@ -25,9 +25,12 @@ def indices_from_labels(obj, labels):
     else:
         labels = labels.astype(obj.index.dtype)
 
-    lhs = cudf.DataFrame({}, index=labels)
+    # There is a chance that the join might jumble the index ordering
+    # so we will sort it with its initial ordering which is stored
+    # in column "__"
```
Comment (kkraus14) [inline python/cudf/cudf/core/indexing.py]:
```suggestion
    # join is not guaranteed to maintain the index ordering
    # so we will sort it with its initial ordering which is stored
    # in column "__"
```

File: python/cudf/cudf/tests/test_indexing.py [hunk 2]
```diff
@@ -889,6 +889,19 @@ def test_out_of_bounds_indexing():
         a[[0, 1, -4]] = 2
 
 
+def test_sliced_indexing():
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_indexing.py]: Could you add a test against a non-default index as well?

**Comment by rgsl888prabhu** [comment]
> > This applies to only if key provided is a array like object and I am not sure about any less expensive alternative.
> 
> It was on my todo list but I dropped the ball. Could you raise an issue with a feature request for libcudf support for something to accomplish the behavior of `loc`? This doesn't include index columns, but just generating a gather map based on values in a specified set of columns.

And the gather map should be in the same order as input if I am not wrong.

File: python/cudf/cudf/tests/test_indexing.py (hunk 2 above)
Comment (rgsl888prabhu) [reply python/cudf/cudf/tests/test_indexing.py]: This a non-default index, I am setting column `a` as index.

**Comment by rgsl888prabhu** [comment]
> > This applies to only if key provided is a array like object and I am not sure about any less expensive alternative.
> 
> It was on my todo list but I dropped the ball. Could you raise an issue with a feature request for libcudf support for something to accomplish the behavior of `loc`? This doesn't include index columns, but just generating a gather map based on values in a specified set of columns.

I have created a Feature request, I am not that good with explaining stuff, so if you find if there are any missing information, please update #4574

**Comment by rgsl888prabhu** [comment]
rerun tests

**Pull request merged**